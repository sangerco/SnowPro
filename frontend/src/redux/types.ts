export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;