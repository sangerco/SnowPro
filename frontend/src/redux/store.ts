import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
