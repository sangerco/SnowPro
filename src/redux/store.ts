import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
