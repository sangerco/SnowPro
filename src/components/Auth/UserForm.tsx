import React, { useEffect, useState } from "react";
import { createUser } from "../../redux/slices/authSlice";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../../redux/store";
import {
  Form,
  Button,
  Grid,
  Segment,
  Card,
  Dimmer,
  Loader,
  Header,
  Divider,
  Rating,
} from "semantic-ui-react";
import { useNavigate } from "react-router";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import { Link } from "react-router-dom";

const UserForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const isAuthenticated = useSelector(
    (state: RootState) => state.auth.isAuthenticated
  );
  const reviews = useSelector((state: RootState) => state.reviews);
  const navigate = useNavigate();

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  const initialState = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
  };

  const [formData, setFormData] = useState(initialState);

  useEffect(() => {
    if (isAuthenticated) {
      navigate("/");
    }
  }, [navigate, isAuthenticated]);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(
      createUser({
        username: formData.username,
        password: formData.password,
        firstName: formData.firstName,
        lastName: formData.lastName,
        email: formData.email,
      })
    );

    setFormData(initialState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  return (
    <Grid>
      <Grid.Row>
        <Grid.Column width={4}>
          {skiAreaState.loading ? (
            <Dimmer active>
              <Loader>Loading...</Loader>
            </Dimmer>
          ) : skiAreaState.error ? (
            <Dimmer active>
              <Header as="h1">
                Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
              </Header>
            </Dimmer>
          ) : skiAreas && skiAreas.length > 0 ? (
            <Card style={{ marginTop: "10px", marginLeft: "20px" }}>
              {skiAreas.map((sa) => (
                <Card.Content key={sa.slug}>
                  <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                </Card.Content>
              ))}
            </Card>
          ) : null}
        </Grid.Column>
        <Grid.Column width={8}>
          <Segment raised>
            <Form onSubmit={handleSubmit} data-testid="user-form">
              <Form.Field required>
                <label htmlFor="username">Username</label>
                <input
                  id="username"
                  placeholder="username"
                  name="username"
                  value={formData.username}
                  onChange={handleChange}
                />
              </Form.Field>
              <Form.Field required>
                <label htmlFor="password">Password</label>
                <input
                  id="password"
                  placeholder="password"
                  type="password"
                  name="password"
                  value={formData.password}
                  onChange={handleChange}
                />
              </Form.Field>
              <Form.Field required>
                <label htmlFor="firstName">First Name</label>
                <input
                  id="firstName"
                  placeholder="First Name"
                  name="firstName"
                  value={formData.firstName}
                  onChange={handleChange}
                />
              </Form.Field>
              <Form.Field required>
                <label htmlFor="lastName">Last Name</label>
                <input
                  id="lastName"
                  placeholder="Last Name"
                  name="lastName"
                  value={formData.lastName}
                  onChange={handleChange}
                />
              </Form.Field>
              <Form.Field required>
                <label htmlFor="email">Email</label>
                <input
                  id="email"
                  placeholder="Email"
                  type="email"
                  name="email"
                  value={formData.email}
                  onChange={handleChange}
                />
              </Form.Field>
              <div className="ui two buttons">
                <Button
                  data-testid="registerButton"
                  color="green"
                  type="submit">
                  Register
                </Button>
                <Button
                  data-testid="cancelButton"
                  color="red"
                  onClick={() => navigate("/")}>
                  Cancel
                </Button>
              </div>
            </Form>
          </Segment>
        </Grid.Column>
        <Grid.Column width={4}>
          {reviews.reviews && reviews.reviews.length > 0 ? (
            <>
              {reviews.reviews.map((review) => (
                <Card>
                  <Card.Content key={review.id} id="review-card">
                    <Card.Header>
                      <Link to={`/ski-areas/reviews/${review.id}`}>
                        {review.header}
                      </Link>{" "}
                      <Divider />
                      {review.skiAreaName}
                    </Card.Header>
                    <Card.Description>By {review.username}</Card.Description>
                    <Card.Meta>
                      <Rating
                        icon="star"
                        defaultRating={review.stars}
                        maxRating={5}
                        disabled
                      />
                    </Card.Meta>
                    {formatDate(review.createdAt)}
                  </Card.Content>
                </Card>
              ))}
            </>
          ) : null}
        </Grid.Column>
      </Grid.Row>
    </Grid>
  );
};

export default UserForm;
