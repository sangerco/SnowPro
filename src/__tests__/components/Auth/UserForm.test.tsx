/* eslint-disable testing-library/no-node-access */
import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import UserForm from "../../../components/Auth/UserForm";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";
import nock from "nock";

describe("test UserForm", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserForm />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UserForm />
        </BrowserRouter>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should contain a form", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserForm />
        </BrowserRouter>
      </Provider>
    );
    const username = screen.getByLabelText("Username");
    const password = screen.getByLabelText("Password");
    const firstName = screen.getByLabelText("First Name");
    const lastName = screen.getByLabelText("Last Name");
    const email = screen.getByLabelText("Email");
    const registerButton = screen.getByTestId("registerButton");
    const cancelButton = screen.getByTestId("cancelButton");
    expect(username).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(firstName).toBeInTheDocument();
    expect(lastName).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(registerButton).toHaveTextContent("Register");
    expect(cancelButton).toHaveTextContent("Cancel");
  });
  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/new-user")
      .reply(201, {
        token: "faketoken",
        user: {
          username: "testuser",
          password: "password",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
        },
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UserForm />
        </BrowserRouter>
      </Provider>
    );
    expect(nock.pendingMocks().length).toBe(2);

    const username = screen.getByLabelText("Username");
    const password = screen.getByLabelText("Password");
    const firstName = screen.getByLabelText("First Name");
    const lastName = screen.getByLabelText("Last Name");
    const email = screen.getByLabelText("Email");

    fireEvent.change(username, { target: { value: "testuser" } });
    fireEvent.change(password, { target: { value: "password" } });
    fireEvent.change(firstName, { target: { value: "Test" } });
    fireEvent.change(lastName, { target: { value: "User" } });
    fireEvent.change(email, { target: { value: "test@user.com" } });

    const userForm = await screen.findByTestId("user-form");
    fireEvent.submit(userForm);

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
