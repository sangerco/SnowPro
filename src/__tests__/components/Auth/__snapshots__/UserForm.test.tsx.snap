// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test UserForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <div
          class="ui active transition visible dimmer"
          style="display: flex;"
        >
          <div
            class="content"
          >
            <div
              class="ui text loader"
            >
              Loading...
            </div>
          </div>
        </div>
      </div>
      <div
        class="eight wide column"
      >
        <div
          class="ui raised segment"
        >
          <form
            class="ui form"
            data-testid="user-form"
          >
            <div
              class="required field"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                id="username"
                name="username"
                placeholder="username"
                value=""
              />
            </div>
            <div
              class="required field"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                id="password"
                name="password"
                placeholder="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="required field"
            >
              <label
                for="firstName"
              >
                First Name
              </label>
              <input
                id="firstName"
                name="firstName"
                placeholder="First Name"
                value=""
              />
            </div>
            <div
              class="required field"
            >
              <label
                for="lastName"
              >
                Last Name
              </label>
              <input
                id="lastName"
                name="lastName"
                placeholder="Last Name"
                value=""
              />
            </div>
            <div
              class="required field"
            >
              <label
                for="email"
              >
                Email
              </label>
              <input
                id="email"
                name="email"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              class="ui two buttons"
            >
              <button
                class="ui green button"
                data-testid="registerButton"
                type="submit"
              >
                Register
              </button>
              <button
                class="ui red button"
                data-testid="cancelButton"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="four wide column"
      />
    </div>
  </div>
</DocumentFragment>
`;
